providers/logingov.go

```diff
@@ -0,0 +1,216 @@
+		return
+	}
+	email = emailData.Email
+	if !emailData.EmailVerified {
+		err = fmt.Errorf("email %s not listed as verified", email)
+		return
+	}
+	return
+}
+
+// Redeem exchanges the OAuth2 authentication token for an ID token
+func (p *LoginGovProvider) Redeem(redirectURL, code string) (s *SessionState, err error) {
+	if code == "" {
+		err = errors.New("missing code")
+		return
+	}
+
+	claims := &jwt.StandardClaims{
+		Issuer:    p.ClientID,
+		Subject:   p.ClientID,
+		Audience:  p.RedeemURL.String(),
+		ExpiresAt: int64(time.Now().Add(time.Duration(5 * time.Minute)).Unix()),
+		Id:        randSeq(32),
+	}
+	token := jwt.NewWithClaims(jwt.GetSigningMethod("RS256"), claims)
+	ss, err := token.SignedString(p.JWTKey)
+	if err != nil {
+		return
+	}
+
+	params := url.Values{}
+	params.Add("client_assertion", ss)
+	params.Add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+	params.Add("code", code)
+	params.Add("grant_type", "authorization_code")
+
+	var req *http.Request
+	req, err = http.NewRequest("POST", p.RedeemURL.String(), bytes.NewBufferString(params.Encode()))
+	if err != nil {
+		return
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	var resp *http.Response
+	resp, err = http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	var body []byte
+	body, err = ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		return
+	}
+
+	if resp.StatusCode != 200 {
+		err = fmt.Errorf("got %d from %q %s", resp.StatusCode, p.RedeemURL.String(), body)
+		return
+	}
+
+	// Get the token from the body that we got from the token endpoint.
+	var jsonResponse struct {
+		AccessToken string `json:"access_token"`
+		IDToken     string `json:"id_token"`
+		TokenType   string `json:"token_type"`
+		ExpiresIn   int64  `json:"expires_in"`
+	}
+	err = json.Unmarshal(body, &jsonResponse)
+	if err != nil {
+		return
+	}
+
+	// XXX should we check signature on JWT and nonce here?
+
+	// Get the email address
+	var email string
+	email, err = emailFromUserInfo(jsonResponse.AccessToken, p.ProfileURL.String())
+	if err != nil {
+		return
+	}
+
+	// Store the data that we found in the session state
+	s = &SessionState{
+		AccessToken: jsonResponse.AccessToken,
+		IDToken:     jsonResponse.IDToken,
+		ExpiresOn:   time.Now().Add(time.Duration(jsonResponse.ExpiresIn) * time.Second).Truncate(time.Second),
+		Email:       email,
+	}
+	return
+}
+
+// GetLoginURL overrides GetLoginURL to add login.gov parameters
+func (p *LoginGovProvider) GetLoginURL(redirectURI, state string) string {
+	var a url.URL
+	a = *p.LoginURL
+	params, _ := url.ParseQuery(a.RawQuery)
+	params.Set("redirect_uri", redirectURI)
+	params.Set("approval_prompt", p.ApprovalPrompt)
+	params.Add("scope", p.Scope)
+	params.Set("client_id", p.ClientID)
+	params.Set("response_type", "code")
+	params.Add("state", state)
+	params.Add("acr_values", p.AcrValues)
+	params.Add("nonce", p.Nonce)
```

Comment:
What is the purpose of this nonce? It seems to only be used here? Is it strictly required?

Reply:
It is required:  https://developers.login.gov/oidc/#authorization , which refers to https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes  Without it and the acr_values param, login.gov will refuse to authorize.

Reply:
FYI, just plugged in a check to make sure that the nonce is proper too.

Reply:
One concern that I do have about the nonce is that it really ought to be done per-session, but the session state is created in OAuthCallback, not OAuthStart when it needs to be created/stored.  It seems like this would take a lot of restructuring to fix, unless you have any great ideas.

Reply:
I can't see a way to easily create the nonce on a per session basis at the moment, but I agree it is something we should probably do in the long term. Could you add a TODO somewhere to signify that we should do this?

Also, as far as I can see, we aren't currently using the nonce when creating the signed JWT, is that correct or have I missed something?

Reply:
To clarify, the nonce is supplied by us in the authorization step, and then it is given back in the signed JWT id_token that we get back from the login.gov IDP during the redeem step.  We do not supply or use the nonce in any other context.  I now have code that checks the signature and the nonce in the id_token, so we do everything we can to make sure that everything is proper.  All that remains is for us to figure out how to supply a unique nonce per session.